### ISSUE TITLE
bug: stack overflow in `TypeChecker`

### ISSUE THREAD

**Issue by xudong963** [datafuselabs/databend#6430]
### Search before asking

- [X] I had searched in the [issues](https://github.com/datafuselabs/databend/issues) and found no similar issues.

### Version

main

### What's Wrong?

`TypeChecker` will cause stack overflow in the case
```sql
set enable_planner_v2 = 1;
select
    o_year,
    sum(case
            when nation = 'BRAZIL' then volume
            else 0
        end) / sum(volume) as mkt_share
from
    (
        select
            extract(year from o_orderdate) as o_year,
            l_extendedprice * (1 - l_discount) as volume,
            n2.n_name as nation
        from
            part,
            supplier,
            lineitem,
            orders,
            customer,
            nation n1,
            nation n2,
            region
        where
                p_partkey = l_partkey
          and s_suppkey = l_suppkey
          and l_orderkey = o_orderkey
          and o_custkey = c_custkey
          and c_nationkey = n1.n_nationkey
          and n1.n_regionkey = r_regionkey
          and r_name = 'AMERICA'
          and s_nationkey = n2.n_nationkey
          and o_orderdate between to_date('1995-01-01') and to_date('1996-12-31')
          and p_type = 'ECONOMY ANODIZED STEEL'
        and 1 and 1 and 1 and 1 and 1
    ) as all_nations
group by
    o_year
order by
    o_year;
```
The bug was found in https://github.com/datafuselabs/databend/runs/7174312042?check_suite_focus=true tpch-q8.

Fortunately, tpch q8 does not stack overflow in main, but it does in the ticket https://github.com/datafuselabs/databend/pull/6412, which is very strange...(This ticket just adds a match branch to the `resolve` function)

So I added some more `and` in `where` to tpch q8, and as expected, the main branch also stack overflowed.

### How to Reproduce?

Run the above case or construct a SQL with many `and` in `where`

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

**Comment by xudong963** [comment]
cc @leiysky

**Comment by sundy-li** [comment]
So we need to rewrite the TypeChecker using visitor instead of recursive call ?

**Comment by xudong963** [comment]
> So we need to rewrite the TypeChecker using visitor instead of recursive call ?

There doesn't seem to be a better way to process the issue. Can you locally reproduce it?

**Comment by sundy-li** [comment]
> > So we need to rewrite the TypeChecker using visitor instead of recursive call ?
> 
> There doesn't seem to be a better way to process the issue. Can you locally reproduce it?

Yes, I can reproduce this with branch main using the SQL you pasted.

**Comment by leiysky** [comment]
I have no idea for this. It seems the only resolution is increasing the stack size in production environment.

It's possible to shrink the state by moving everything to heap to reduce stack usage, but it needs time for evaluation. Since the type check is mutual recursion with subquery binding, the biggest memory consumer is unrecognizable for complicated queries.

**Comment by xudong963** [comment]
Maybe we can use https://docs.rs/stacker/latest/stacker/ to check stack usage and grow the stack space during recursion.

**Comment by leiysky** [comment]
> Maybe we can use https://docs.rs/stacker/latest/stacker/ to check stack usage and grow the stack space during recursion.

Not exactly, the stack usage of async function is complicated, I'm writing a PR to do some straightforward optimization.
